### PR TITLE
Allow Angle argument to FlatPlate dictionary for SetSample

### LINKED ISSUE #17194: Allow for rotation in flat plate in SetSample
[SetSample](http://docs.mantidproject.org/nightly/algorithms/SetSample-v1.html#id5) allows for a flat plate geometry but it is currently constrained so that it's surface normal is parallel to the beam.

Add an optional `Angle` parameter to the `FlatPlate` configuration defining it as the angle, in degrees, between the perpendicular and the beam, with the default=0.0 in line with the current behaviour.

### PR DESCRIPTION
**Pull request by martyngigg**
A flat plate sample shape can now be rotated about its own center using `SetSample`

**To test:**

The following script produces a plot of the attenutation factor at a fixed wavelength versus the angle of rotation of the flat plate.

```python
from mantid.simpleapi import *
from matplotlib import pyplot as plt
import numpy as np

data = CreateSampleWorkspace(WorkspaceType='Histogram', NumBanks=1)
data = CropWorkspace(data, StartWorkspaceIndex=0, EndWorkspaceIndex=0)
data = ConvertUnits(data, Target="Wavelength")
SetSample(data,  Material={'ChemicalFormula': 'V', 'SampleNumberDensity': 0.073})

# geometry in cm
sample_height = 2.0
sample_width = 1.0
sample_thick = 0.1

# simulation options
nlambda_cyl = 20
nlambda_mc = 20
nevents = 20000
interpolation = 'Linear'

# Compute absorption for 
angles = np.arange(91., step=5.)
lambda_index = 50
abs_factors = []
for angle in angles:
    SetSample(data, Geometry={'Shape': 'FlatPlate', 'Height': sample_height, 'Width': sample_width,
                                'Thick': sample_thick, 'Center': [0.,0.,0.], 'Angle': float(angle)})
    abs = MonteCarloAbsorption(data,
                                                  NumberOfWavelengthPoints=nlambda_mc,
                                                  EventsPerPoint=nevents)
    abs_factors.append(abs.readY(0)[lambda_index])
    DeleteWorkspace(abs)
#endfor

# Plot absorption factor for single wavelength against the angle of rotation
plt.plot(angles, abs_factors)
plt.show()
```

As the sample rotates the width and thickness dimensions swap and the attenuation increases due to the extra distance travelled. This manifests itself as a reduction in the fraction of the beam that makes it through:

flate_plate_abs_vs_angle_step_5

Fixes #17194 

[RELEASE NOTES](https://github.com/mantidproject/mantid/commit/a6724b0f17c7d22e17004f701c47db2a10a8dbeb)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [x] Is user-facing documentation written in a user-friendly manner?
- [x] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/inc/MantidDataHandling/SetSample.h [hunk 1]
```diff
@@ -49,13 +50,18 @@ class MANTID_DATAHANDLING_DLL SetSample final : public API::Algorithm {
 
   const Geometry::SampleEnvironment *
   setSampleEnvironment(API::MatrixWorkspace_sptr &workspace,
-                       const Kernel::PropertyManager &args);
+                       const Kernel::PropertyManager_sptr &args);
```
Comment (soininen) [inline Framework/DataHandling/inc/MantidDataHandling/SetSample.h]: I'm wondering why pass shared pointers as references? Also, would it be possible to use `PropertyManager_const_sptr` here? Only some getters are used in the method as far as I can see. Same for the `setSampleShape` method and maybe others in this class.

File: Framework/DataHandling/src/SetSample.cpp [hunk 2]
```diff
@@ -258,10 +243,13 @@ void SetSample::setSampleShape(API::MatrixWorkspace_sptr &workspace,
   */
 
   // Try known shapes or CSG first if supplied
-  const auto xml = tryCreateXMLFromArgsOnly(args);
-  if (!xml.empty()) {
-    runSetSampleShape(workspace, xml);
-    return;
+  if (args) {
+    auto refFrame = workspace->getInstrument()->getReferenceFrame();
```
Comment (soininen) [inline Framework/DataHandling/src/SetSample.cpp]: I think `refFrame` could be delcared `const`.

File: docs/source/algorithms/SetSample-v1.rst [hunk 3]
```diff
@@ -71,14 +71,20 @@ For defining the full shape a key called ``Shape`` specifying the desired shape
 expected along with additional keys specifying the values (all values are assumed to
```
Comment (soininen) [inline docs/source/algorithms/SetSample-v1.rst]: While at it, you could perhaps fix the typo in the first sentence of the **Environment** subsection as well: 'enviorment' -> 'environment'.

File: docs/source/algorithms/SetSample-v1.rst [hunk 4]
```diff
@@ -132,11 +138,11 @@ The following example uses a test file called ``CRYO-01.xml`` in the
 
    # A fake host workspace, replace this with your real one.
    ws = CreateSampleWorkspace()
-   # Use geometry from environment but set differnet height for sample
+   # Use geometry from environment but set different height for sample
    SetSample(ws, Environment={'Name': 'CRYO-01', 'Container': '8mm'},
              Geometry={'Shape': 'HollowCylinder', 'Height': 4.0,
                        'InnerRadius': 0.8, 'OuterRadius': 1.0,
-                       'Center': [0.,0.,0.], 'Axis':1},
```
Comment (soininen) [inline docs/source/algorithms/SetSample-v1.rst]: There is a space missing between the colon and '1'.

**Review by soininen — CHANGES_REQUESTED** [review]

**Comment by soininen** [comment]
Are there any plans to extend `SetSample::validateInputs()`? The algorithm seems to happily accept negative sample dimensions.

File: Framework/DataHandling/inc/MantidDataHandling/SetSample.h (hunk 1 above)
Comment (martyngigg) [inline Framework/DataHandling/inc/MantidDataHandling/SetSample.h]: Passing the `shared_ptr`'s by const reference here just avoids increasing the reference count: http://stackoverflow.com/a/37610721. I know I have a reference to keep the object alive at the top level as this is all private. I would usually prefer a plain `const PropertyManager&` but it needs to be a `shared_ptr` to be fed into the child algorithm.  

You're right I can probably use `PropertyManager_const_sptr` here.

File: Framework/DataHandling/src/SetSample.cpp (hunk 2 above)
Comment (martyngigg) [inline Framework/DataHandling/src/SetSample.cpp]: :+1:

File: docs/source/algorithms/SetSample-v1.rst (hunk 3 above)
Comment (martyngigg) [inline docs/source/algorithms/SetSample-v1.rst]: :+1:

**Comment by martyngigg** [comment]
You're right, I should provide better checking for the basic cases such as negative dimensions. I'll update the code.

**Comment by martyngigg** [comment]
@soininen I've added some changes after your feedback.

File: Framework/DataHandling/src/SetSample.cpp [hunk 5]
```diff
@@ -94,28 +143,61 @@ const std::string SetSample::summary() const {
...
   std::map<std::string, std::string> errors;
 
+  auto existsAndNotEmptyString =
+      [](const PropertyManager &pm, const std::string &name) {
+        if (pm.existsProperty(name)) {
+          const auto value = pm.getPropertyValue(name);
+          if (value.empty())
+            return false;
+          else
+            return true;
```
Comment (soininen) [inline Framework/DataHandling/src/SetSample.cpp]: Couldn't you just `return !value.empty();`?

File: Framework/DataHandling/src/SetSample.cpp [hunk 6]
```diff
@@ -94,28 +143,61 @@ const std::string SetSample::summary() const {
...
-      std::string name = environArgs->getPropertyValue("Container");
-      if (name.empty()) {
-        errors["Environment"] = "Environment 'Can' flag is an empty string!";
+    if (!existsAndNotEmptyString(*environArgs, SEArgs::CONTAINER)) {
+      errors[PropertyNames::ENVIRONMENT] =
+          "Environment flags require a non-empty 'Container' entry.";
+    }
+  }
+
+  //  // Validate as much of the shape information as possible
```
Comment (soininen) [inline Framework/DataHandling/src/SetSample.cpp]: A bit nit-picking, but perhaps the comment needs some cleanup.

**Review by soininen — CHANGES_REQUESTED** [review]

File: Framework/DataHandling/src/SetSample.cpp (hunk 5 above)
Comment (martyngigg) [inline Framework/DataHandling/src/SetSample.cpp]: Good point :man_facepalming:

**Review by soininen — APPROVED** [review]

**Comment by soininen** [comment]
The new angle property seems to work as expected. The test script produces the expected cyclic pattern when the sample is rotated around [-180, 450] degrees. The input properties are validated more rigorously than before, and there are new units tests in place. I think we are done here :shipit:

**Pull request merged**